support presentation.panel in task config

elaihau opened the pull request:
- updated the task config schema to support controlling whether the terminal instance is shared between task runs.

Signed-off-by: Liang Huang <lhuang4@ualberta.ca>


#### How to test

1) Open a `tasks.json`, when adding properties to a `task.presentation`, users should get intellisense for the new property `task.presentation.panel`.

2) Without `presentation.panel` being added to any task configs, Theia should reuse task terminals as much as possible. 
  - If a terminal is idle (i.e., with no tasks running), it should be considered as a candidate to run the next task
  - if a terminal is busy (i.e., has a running task), it should not be considered as a candidate to run the next task, until the running task finishes.

![Peek 2020-03-02 22-41](https://user-images.githubusercontent.com/37082801/75741676-50b63b00-5cd9-11ea-98cc-d3cc5adb808e.gif)



3) With `presentation.panel === 'dedicated'` added to a task config, a terminal should be reserved for this particular task. Other tasks will not be started in this "reserved terminal".

4) With `presentation.panel === 'new'` added to a task config, a new terminal should be opened for this task, every single time. However, the new terminals are not reserved for this particular task. In other words, they will be considered as candidates to run other tasks (whose `presentation.panel === 'shared' or not defined`).

![Peek 2020-03-02 22-43](https://user-images.githubusercontent.com/37082801/75741988-45afda80-5cda-11ea-82ed-84e4e643d8da.gif)

5) Once a task finishes or is terminated, a line of bold text `Terminal will be reused by tasks` should show up in the terminal. Unlike what vs code does, we don't close the terminal on receiving a keyboard event.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

lmcbout commented:
I have a question. 
When there are 3 terminals available, the focus is on the third one. Then you initiate a task, should the task run in the terminal having he focus or the first terminal open in the workspace?
Here my task.presentation
 "presentation": {
              "reveal": "always",
              "focus": true,
              "panel": "shared"
            }

vince-fugnitto commented:
> I have a question.
> When there are 3 terminals available, the focus is on the third one. Then you initiate a task, should the task run in the terminal having he focus or the first terminal open in the workspace?
> Here my task.presentation
> "presentation": {
> "reveal": "always",
> "focus": true,
> "panel": "shared"
> }

The current implementation chooses the first `idle` terminal to run the task in.
We could improve the check to first look for an active `idle` terminal, and if not check for just `idle`, but I'm fine with either approach.

lmcbout commented:
>The current implementation chooses the first idle terminal to run the task in.
We could improve the check to first look for an active idle terminal, and if not check for just idle, but I'm fine with either approach.

Personnaly, I would go for the active idle terminal first, so the end-user sees the task running, it is not hidden in another terminal. If the focus is elsewhere, not on any terminal when the task is initiated, then the first  empty terminal gets the focus to show the task running is fine. I would not block the merge for it.

 Tested some test cases and seems to work fine.

elaihau commented:
> Personnaly, I would go for the active idle terminal first, so the end-user sees the task running, it is not hidden in another terminal. If the focus is elsewhere, not on any terminal when the task is initiated, then the first empty terminal gets the focus to show the task running is fine. I would not block the merge for it.

i agree with you and Vincent. running the task in the active idle terminal is more user friendly. i will update the code

elaihau commented:
@lmcbout @vince-fugnitto 
i made it work the way you suggested however i am not sure if i did it the right way. please see my comments for details.

![Peek 2020-03-03 20-56](https://user-images.githubusercontent.com/37082801/75837088-d1d10900-5d91-11ea-8a89-8d1f7a91f186.gif)

elaihau commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -185,10 +185,10 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
+    protected updateCurrentTerminal(widget?: TerminalWidget): void {
+        const currentWidget = widget || this.shell.currentWidget;
+        if (currentWidget instanceof TerminalWidget) {
+            this.setCurrentTerminal(currentWidget);
```
i made this change because i found that
if we do 
```
const newWidget = widgetManager.getOrCreateWidget(ID, options);
this.applicationShell.revealWidget(newWidget.id);
```
`this.shell.currentWidget` is the most recent widget that I clicked, not the terminal widget I just created.

therefore, in my opinion, in the context of `TerminalFrontendContribution` class, I made the `currentTerminal` the one that we opened (i.e.., activate or reveal) most recently.

What do you think?
@vince-fugnitto @RomanNikitenko @akosyakov @lmcbout

akosyakov replied:
please revert, current is last focused, reveal does not change focus, you should use  `activate` if you want to give a focus to the terminal

akosyakov replied:
> this.shell.currentWidget is the most recent widget that I clicked, not the terminal widget I just created.

You should use `TerminalFrontendContribution.currenTerminal`.

> therefore, in my opinion, in the context of TerminalFrontendContribution class, I probably should make the currentTerminal the one that we opened (i.e.., activate or reveal) most recently.

Semantics of `TerminalFrontendContribution.currenTerminal` is correct and cannot be changed. It is a currently focused terminal or one which most recently had focused. Not anything else.

elaihau commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -500,6 +500,7 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
         } else if (op.mode === 'reveal') {
             this.shell.revealWidget(widget.id);
         }
+        this.updateCurrentTerminal(widget);
```
this line too. 
if we open a terminal widget, we consider it as the `current terminal`.

akosyakov replied:
please revert, current means it was focused, it can only happen if `activate` was passed and you should await here for activation to be finished that clients can be sure about it

Any other logic is bogus. The single source of truth of focus state is the application shell, everything should be derived from there.

elaihau replied:
Got it. I will add another function to the `TerminalService` to keep track of the last `actived/revealed terminal`

akosyakov replied:
Why do you need to know last revealed?

elaihau replied:
because i want the task to run in the terminal widget that "visibly" stays at the top, and the `TerminalFrontendContribution.currenTerminal` is not always the one.

For example, as a user you can set focus to TerminalA, and after that, TerminalB is revealed. In this case, B stays on top of A. While `TerminalFrontendContribution.currenTerminal` still points to A, I want to run the next task in B.

lmcbout commented:
I have another question:
If I open a Terminal and then start a TASK, it will never used the Terminal created from the button: "New Terminal". We only re-used the Terminal created from a Task. Wether I set or not the presentation arguments. If I put the focus on the created Terminal (Not the one created from the Task) and start a Task, still don't use the manual created Terminal. Is it suposed to be like that?

lmcbout commented:
I talk with @vince-fugnitto and for the Task terminal, they are dedicated Terminal that we re-used and the Termainal created manually are not exactly the same type of Terminal.

vince-fugnitto commented:
> I have another question:
> If I open a Terminal and then start a TASK, it will never used the Terminal created from the button: "New Terminal". We only re-used the Terminal created from a Task. Wether I set or not the presentation arguments. If I put the focus on the created Terminal (Not the one created from the Task) and start a Task, still don't use the manual created Terminal. Is it suposed to be like that?

It works as intended. These terminals are dedicated to tasks and are not user-triggered integrated terminals. VS Code does not even allow such terminals to be re-used and only prompts users to close them (not re-use):

![image](https://user-images.githubusercontent.com/40359487/75881811-cf9c9800-5ded-11ea-80a2-c171bbb90413.png)

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,126 @@
+
+        // we are unable to find a terminal widget to run the task, or `taskPresentation === 'new'`
+        if (!reusableTerminalWidget) {
+            const widget = <TerminalWidget>await this.widgetManager.getOrCreateWidget(TERMINAL_WIDGET_FACTORY_ID, options);
```
`TERMINAL_WIDGET_FACTORY_ID` should implementation details of `TerminalService`, maybe you are better to use `TerminalService.newTerminal` and get rid of dependencies to `WidgetManager`

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,126 @@
+    protected readonly terminalService: TerminalService;
+
+    // Map indexed by terminal widget id
+    protected terminalWidgetMap: Map<string, { isDedicated: boolean, taskConfig?: TaskConfiguration, widget: TerminalWidget }> = new Map();
```
It does not look good, we should not cache `TerminalWidget` instances, it will lead to memory leaks. `WidgetManager` should be used as a single source of the truth. In this case you should use `TerminalService` as a facade to it.

You can cache though terminal widget ids as a reference which can be resolved later against `TerminalService` (and `WidgetManager` eventually)

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -25,6 +25,11 @@ export abstract class TerminalWidget extends BaseWidget {
+    /**
+     * Whether the terminal widget runs a task
+     */
+    abstract hasRunningTask: boolean;
```
please revert, terminal should not know anything about tasks

elaihau replied:
could you please suggest another way to check if the terminal is idle?
or maybe i can simply rename `hasRunningTask` to `isIdle` to make it look irrelevant to tasks?

akosyakov replied:
How it would be useful if we forget about tasks? There should be conceptual explanation.

elaihau replied:
Yes `isIdle` is conceptually irrelevant to tasks.
in the terminal widget class `isIdle` is set `true` once the attached process finishes.

akosyakov replied:
I mean what does it mean?

akosyakov commented on packages/terminal/src/node/terminal-backend-contribution.ts:
```diff
@@ -39,6 +39,9 @@ export class TerminalBackendContribution implements MessagingService.Contributio
                 connection.onClose(() => output.dispose());
                 connection.listen();
+                termProcess.onClose(() => {
+                    connection.sendNotification('onClose', undefined);
```
Why do we need it?

elaihau replied:
this gives the front-end widget the opportunity to know the task has finished.

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -402,6 +412,11 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
 
+                connection.onNotification('onClose', event => {
+                    this.isIdle = true;
+                    this.term.writeln('\x1b[1m\n\rTerminal will be reused by tasks. \x1b[0m\n');
```
Terminals should know nothing about tasks. It breaks abstractions, such code should belong to tasks extensions.

You don't need `onClose` notification, there is already `connection.onDispose` and `connection.onClose` on the json-rpc level.

elaihau commented:
@akosyakov 
I redefined some interfaces and move the task-specific implementation to the `task` extension, in order to not break the abstraction in `terminal`.

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -189,11 +189,39 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
+        const lastUsedTerminalId = lastUsed.id;
+        if (!this.lastUsedTerminal || this.getLastUsedTerminalId() !== lastUsedTerminalId) {
+            this.mostRecentlyUsedTerminalIds.push(lastUsedTerminalId);
+            lastUsed.onTerminalDidClose(() => {
```
maybe `onDispose`? i.e. if a terminal process was not ever opened we still should clean up `mostRecentlyUsedTerminalIds` then a widget is disposed

elaihau replied:
done.

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -189,11 +189,39 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
         const widget = this.shell.currentWidget;
         if (widget instanceof TerminalWidget) {
             this.setCurrentTerminal(widget);
+            this.setLastUsedTerminal(widget);
```
Should not it go into `setCurrentTerminal`?

elaihau replied:
Two differences between `setCurrentTerminal()` and `setLastUsedTerminal()`:

1) `setCurrentTerminal()` can be called with `undefined`, i.e., `this.setCurrentTerminal(undefined)`, meaning the "current terminal" is not any of the opened terminal widgets.
`setLastUsedTerminal()` won't be called with `undefined`.

2) `setLastUsedTerminal()` can be called with terminal widget that does not have focus, while `setCurrentTerminal()` (except for `undefined`) is always called with a widget that has the focus.

So i did not put the `setCurrentTerminal` calls into `setCurrentTerminal`

akosyakov replied:
ok

akosyakov commented:
@elaihau i've looked only at the terminal extension and it looks clean now, thank you!

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,166 @@
+        });
+    }
+
+    async getOrCreateWidget(options: TaskTerminalWidgetOpenerOptions): Promise<TerminalWidget> {
```
I would call it `open` since it does not just create a terminal, but actually opens it, also ti accepts opener options.

akosyakov commented:
@elaihau code wise it looks clean, but you need to find someone to test it. cc @RomanNikitenko ?

RomanNikitenko commented:
@elaihau 
I'm able to test the changes in the evening.
Please feel free to not wait my review if you need to get the PR merged sooner.

thank you!

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -495,10 +523,13 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
+            } else {
+                this.shell.revealWidget(widget.id);
+            }
+            this.setLastUsedTerminal(widget);
```
I am not sure that's a correct way to set it, someone could call `shell.revealWidget` for a terminal from other code path and we will miss it. Maybe we should rely on `onDidChangeVisibility` event `isHidden` flag instead?

elaihau replied:
Let me try. My first guess is that ‘revealWidget’ does not mutate the visibility.

akosyakov replied:
It’s the purpose of revealwidget.

elaihau replied:
I implemented what you proposed. You are right - it works. Thank you for the suggestion !

elaihau commented:
> So a restored terminal is never used.
> I guess the cause of the issue is: https://github.com/eclipse-theia/theia/pull/7260/files#diff-38e65dfb04ff1bf292adf8abf4e1b9d5R60

Thank you very much for the review @RomanNikitenko !
Could you please point me to the line you referred in your comment one more time ? I force-pushed my latest change before clicking your link (which became invalid after the force-push) 🙈

My apologies.

RomanNikitenko commented:
@elaihau 
> Could you please point me to the line you referred in your comment one more time ?

I mean we have the issue with page refreshing because:
- the restored terminal widget is never put to `terminalWidgetMap` and can not be reused.
- `terminalWidgetMap` is empty after refresh. 
As result the `dedicated` terminal widget can not be found for the case described above and a new terminal widget is created for next running. 

I didn't check it - just my thoughts, sorry if I misled you.

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -411,7 +411,6 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
         this.toDisposeOnConnect.dispose();
         this.toDispose.push(this.toDisposeOnConnect);
-        this.term.reset();
```
I think it was important for some reasons,  like you please track back while it was added. We have to test for regressions.

elaihau replied:
yes it is important. in my pr it is moved to this place https://github.com/eclipse-theia/theia/pull/7260/files#diff-38e65dfb04ff1bf292adf8abf4e1b9d5R99

akosyakov replied:
but it is not about tasks, it's about resetting a state for terminal generally on reconnection

I think it has to stay here.

akosyakov replied:
Removing it will cause https://github.com/eclipse-theia/theia/issues/2554 again.

elaihau replied:
hmm... please could you advise on how to resolve this, as i am stuck:

1.  `this.term.reset()` clears all output in terminal widget.
2. terminal should have no knowledge of tasks. 
3. we want to retain the existing terminal output when a task is started

so if i put `this.term.reset()` back, i wouldn't be able to achieve # 3 due to the reason described in # 2 .

akosyakov replied:
> `this.term.reset()` clears all output in terminal widget.

It does not clear the output, it resets it from the shell. If the shell is still have output it should be there. What kind of output you mean?

elaihau replied:
I take a closer look at the behavior:

![Peek 2020-03-13 07-51](https://user-images.githubusercontent.com/37082801/76618727-a2439e80-64ff-11ea-892f-a0de67b25fd6.gif)

elaihau replied:
maybe i am wrong. looks like `term.reset()` does not clear the output directly. it is something else that clears the terminal output after the `term.reset()`.

elaihau replied:
ok i believe `terminal.reset()` clears the  terminal itself:

reset() prints '\x1bc':
https://github.com/xtermjs/xterm.js/blob/master/typings/xterm.d.ts#L1007

and '\x1bc' clears the terminal:
https://askubuntu.com/a/616773

and i personally don't think `term.reset()` belongs to `terminal.connectTerminalProcess()`.

in case that you disagree with me, i will find a work-around.

elaihau replied:
@akosyakov 
I was trying to reproduce #2554 without `term.reset()` and I wasn't able to (please see GIF below). Please kindly let me know if i was reproducing it the correct way. Thank you.

![Peek 2020-03-15 19-02](https://user-images.githubusercontent.com/37082801/76712456-b221d500-66ef-11ea-8a77-38f9c04623c3.gif)

akosyakov replied:
I was trying to reproduce it with your PR as well and could not, something seems to change in the meantime. Let's try without it. 👍

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -510,6 +525,8 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
             this.onTermDidClose.fire(this);
             this.onTermDidClose.dispose();
         }
+        this.onTermDidDispose.fire(undefined);
```
are you sure that base class does not have an event or signal already?

elaihau replied:
I checked `BaseWidget` (in theia) and `Widget` (in phorsphor), the closet event that I can find is `Widget.onCloseRequest()`. Is it the right one to use?

akosyakov replied:
ok, let's go with `onTermDidDispose` for now

akosyakov replied:
I thought all widgets have `onDispose` event already

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -194,6 +197,58 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
+            this.mostRecentlyUsedTerminalEntries.splice(entryIndex, 1);
+        } else {
+            toDispose = new DisposableCollection();
+            toDispose.push(
```
I would expect that we hook into ondidcreate hook of WidgetManager and install a listener where for each terminal always.

akosyakov replied:
Nvm I see that you do it already )

elaihau commented:
> My reference was for `terminalWidgetMap`.
> I mean we have the issue with page refreshing because:
> 
> * the restored terminal widget is never put to `terminalWidgetMap` and can not be reused.
> * `terminalWidgetMap` is empty after refresh.
> 
> As result the `dedicated` terminal widget can not be found for the case [described above](https://github.com/eclipse-theia/theia/pull/7260#pullrequestreview-373175001) and a new terminal widget is created for next running.

@RomanNikitenko 
Yes you are right - when browser was refreshed we lost data in the `terminalWidgetMap`.

I did some investigation by adding some code, so that the `TaskTerminalWidgetManager` reacts to the events of terminal recreations (after refresh), and adds the recreated terminals into the `terminalWidgetMap`. I found making that work requires quite a bit of work. Please give me some time to implement the solution. Thank you for your patience in advance.

RomanNikitenko commented:
@elaihau 
As I mentioned above in general it works well for me and it's nice improvement for tasks!
I think mentioned use case is not critical and not blocking user for using task related functionality.
So it's OK for me to improve the behavior in a separate PR, that's why I approved the PR.

Up to you and other reviewers.
thanks!

elaihau commented:
> As I mentioned above in general it works well for me and it's nice improvement for tasks!
> I think mentioned use case is not critical and not blocking user for using task related functionality.
> So it's OK for me to improve the behavior in a separate PR, that's why I approved the PR.

@RomanNikitenko 
I find the solution :)

![Peek 2020-03-16 00-08](https://user-images.githubusercontent.com/37082801/76722597-a7c90080-671a-11ea-8516-e700a888df40.gif)

elaihau commented:
I tested this change one more time after a rebase.

I will merge it in a day if there are no objections.

elaihau commented:
> I tested a little after update and found that it uses my current terminal for running a task.
> Is it expected behavior?
> 

no it is not expected. let me fix it  :)

elaihau commented:
@RomanNikitenko 
could you please check the latest patch when you get the chance?
it should have been fixed.
the tasks wouldn't be started in interactive terminals.

![Peek 2020-03-17 20-26](https://user-images.githubusercontent.com/37082801/76913574-c0a5f300-688d-11ea-8694-bd11db139e14.gif)

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -25,6 +25,9 @@ export abstract class TerminalWidget extends BaseWidget {
 
+    abstract terminalId: number;
+
+    abstract isInteractive: boolean;
```
could you document what it means?

akosyakov replied:
It should be also `readonly interactive: boolean`

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
 export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget {
 
     private readonly TERMINAL = 'Terminal';
+    isInteractive = true;
```
exposed to clients?

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
-    protected terminalId = -1;
+    protected readonly onTermDidDispose = new Emitter<void>();
+    protected readonly onDidReconnect = new Emitter<TerminalWidget>();
     protected fitAddon: FitAddon;
```
why to expose to clients?

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
+    isInteractive = true;
+    terminalId = -1;
     protected readonly onTermDidClose = new Emitter<TerminalWidget>();
-    protected terminalId = -1;
```
Why to expose to clients?

elaihau replied:
because we exposed the terminalId in TaskInfo interface, i.e.,  `TaskInfo.terminalId: number`. And we need it from the terminalWidget to find the association between the running tasks and widgets after the browser is refreshed.

https://github.com/eclipse-theia/theia/pull/7260/files#diff-38e65dfb04ff1bf292adf8abf4e1b9d5R92-R96

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -45,20 +48,29 @@ export abstract class TerminalWidget extends BaseWidget {
 
+    abstract onTerminalDidDispose: Event<void>;
+
+    abstract onDidReconnectTerminalProcess: Event<TerminalWidget>;
```
why don't have `onDidConnect` which fires whenever a connection is established? Is it an event that `isInteracitve` flag changed?

elaihau replied:
yep. my `onDidReconnectTerminalProcess` does the job of `onDidConnect` in your comment. I should rename it.

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -25,6 +25,9 @@ export abstract class TerminalWidget extends BaseWidget {
 
     abstract processId: Promise<number>;
 
+    abstract terminalId: number;
```
`readonly`, clients cannot change it

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
+    terminalId = -1;
     protected readonly onTermDidClose = new Emitter<TerminalWidget>();
-    protected terminalId = -1;
+    protected readonly onTermDidDispose = new Emitter<void>();
```
When do you dispose new emitters? Please use `this.toDispose.push`.

elaihau replied:
i disposed them in the `dispose()` method. and you are right, it is cleaner and more organized to use toDispose.push()

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
     protected readonly onTermDidClose = new Emitter<TerminalWidget>();
-    protected terminalId = -1;
+    protected readonly onTermDidDispose = new Emitter<void>();
+    protected readonly onDidReconnect = new Emitter<TerminalWidget>();
```
please give proper names `onDidReconnectEmitter`, `onDidReconnect` is a name for event, not an emitter

elaihau commented:
@akosyakov I resolved the comments. Please kindly let me know if there are more. Thank you for the review !

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -497,7 +552,7 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
         if (op.mode === 'activate') {
             this.shell.activateWidget(widget.id);
-        } else if (op.mode === 'reveal') {
+        } else {
```
I don't think it is correct, there is a 3rd option `open` which means that a widget should be attached but not revealed. Now it will reveal as well.

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
 export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget {
 
     private readonly TERMINAL = 'Terminal';
+    isInteractiveTerminal = true;
```
`protected _interactive = true`

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
 
     private readonly TERMINAL = 'Terminal';
+    isInteractiveTerminal = true;
+    termId = -1;
```
`protected _terminalId = -1;`

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -215,6 +219,14 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+        return this.termId;
+    }
+
+    get isInteractive(): boolean {
```
`get interactive`

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -432,6 +443,8 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
                 connection.listen();
                 if (waitForConnection) {
                     waitForConnection.resolve(connection);
+                    this.isInteractiveTerminal = this.restored;
```
I don't understand it. `restored` means that a widget was restored on reload, why `interactive` should be assigned to it?

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -432,6 +443,8 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
                 if (waitForConnection) {
                     waitForConnection.resolve(connection);
+                    this.isInteractiveTerminal = this.restored;
+                    this.onTermDidConnectProcessEmitter.fire(this);
```
If this event is supposed to be fired when `interfacive` state changed when it should be called `onDidChangeInteractive`.

Or rename `interactive` to `connected`  and  rename an event `onDidChangeConnected`.

please never shorten, it our coding rule: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#whole-words-names
But don't write over verbose names too, it is clear from `TerminalWidget` that events about terminal there is no need to add `Terminal` to each event name.

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,11 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
+    termId = -1;
     protected readonly onTermDidClose = new Emitter<TerminalWidget>();
-    protected terminalId = -1;
+    protected readonly onTermDidDisposeEmitter = new Emitter<void>();
```
I would prefer if we add `onDidDisposeEmitter` to `BaseWidget` instead that it is available for each widget. Instead of inventing it each time.

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -25,6 +25,11 @@ export abstract class TerminalWidget extends BaseWidget {
 
+    abstract readonly terminalId: number;
+
+    /** A boolean that indicates if the terminal is interactive */
```
js-doc should explain what interactive means without using this word

akosyakov commented:
It is still not clear to me what interactive means and why it is implemented based on restored also there is no event indicating what interactive did change and there is an did connect event without state.

elaihau commented:
> It is still not clear to me what interactive means and why it is implemented based on restored also there is no event indicating what interactive did change and there is a did connect event without state.
> 
> I don't think that we can possible know whether terminal is interactive or not, for reasoning: [#7323 (comment)](https://github.com/eclipse-theia/theia/issues/7323#issuecomment-601056334)

sorry for the confusion.
`interactive` means a terminal that has a shell where a user can key-in commands and find output of the commands, like the one shown in the screenshot below

![Screenshot from 2020-03-19 07-18-33](https://user-images.githubusercontent.com/37082801/77062488-563aa300-69b2-11ea-990e-170930031f77.png)

please advise a better name for this property. thanks !

akosyakov commented:
@elaihau I was suggesting following terminal API:
```ts
export interface TerminalWidgetOptions {
   /** defaults to user */
   readonly kind?: 'user' | string;
}
export interface TeminalWidget {
   readonly kind: 'user' | string;
}
```

`kind` indicates whether a terminal is created by a user or by some extension for a user, e.g. the task extension can provide `task` instead. In such way you look only for terminals with `task` kind to find out terminals to reuse for tasks.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,203 @@
+    protected readonly workspaceService: WorkspaceService;
+
+    // Map indexed by terminal widget id
+    protected terminalWidgetMap: Map<string, { isDedicated: boolean, taskId: number, taskConfig?: TaskConfiguration, isIdle: boolean }> = new Map();
```
It feels like properties of task terminal. I wonder why don't replace it with `TaskTerminalWidget` namespace, i.e.:
```ts
export interface TaskTerminalWidget extends TerminalWidget {
    readonly kind: 'task'
    dedicated?: boolean
    taskId?: boolean
    taskConfig?: TaskConfiguration
    busy?: boolean
}
export namespace TaskTerminalWidget {
    export function is(terminalWidget: TerminalWidget): TerminalWidget is TaskTerminalWidget {
        return terminalWidget.kind === ' task';
    }
}
```

Please also use `isDedicated` for method names, properties should be named `dedicated`. `idle` is fine, but in other places we use `busy`.

elaihau replied:
the "terminal kind" can only be "user" and "other". If we want to have "user", "task", and more, we will need to store it in the local strage when browser refreshes.

akosyakov replied:
yes, It will be stored since it belongs to `TerminalWidgetOptions`.

elaihau commented:
> @elaihau
> Tested after update and I still can reproduce both issues:
> 
> * related to page refreshing
>
> * related to using current terminal for running a task
> 
> I've double-checked that I have your last changes.
> Could you check it on your side?
> Can you reproduce these cases?

@RomanNikitenko 
my bad. I broken something when I refactored the code. 
I retested both scenarios you described. They should work now.

@akosyakov 
I implemented your proposal.

Thank both of you for your time, really appreciated !

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -47,8 +47,9 @@ export interface TerminalWidgetFactoryOptions extends Partial<TerminalWidgetOpti
 
     private readonly TERMINAL = 'Terminal';
+    protected terminalKind = 'user';
+    protected termId = -1;
```
please don't shorten but rename it to `_terminalId`, see https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#underscore_accessors

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -325,6 +340,7 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         if (IBaseTerminalServer.validateId(terminalId)) {
             return terminalId;
         }
+        this.terminalKind = 'user';
```
I don't think we should change a value of kind after construction, it should be assigned once and be always readonly.

What was your motivation?

elaihau replied:
the reason behind this line is, when a terminal fails to attach to a process, it becomes a "user terminal". Please see this GIF:

![Peek 2020-03-23 18-03](https://user-images.githubusercontent.com/37082801/77367562-cfc6ee00-6d30-11ea-9d9d-3e6fe3e25174.gif)

akosyakov replied:
> the reason behind this line is, when a terminal fails to attach to a process, it becomes a "user terminal".

I don't think it is correct from the lifecycle and responsibility perspective. If the task extensions is responsible for creating a task terminal, then it should be responsible for destroying it or giving up the ownership (i.e. changing a kind of terminal to `user`), not a terminal itself.

Why don't just print `Terminal will be reused for tasks`? It removes new logic with the managing terminal ownership after creation and is more transparent for a user.

elaihau replied:
> Why don't just print `Terminal will be reused for tasks`? It removes new logic with the managing terminal ownership after creation and is more transparent for a user.

this is in the terminal extension. since terminals have no knowledge of task, i guess we probably need to add another event to inform the task extension that this terminal widget fails to attach to a process ...

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -406,15 +422,14 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         const waitForConnection = this.waitForConnection = new Deferred<MessageConnection>();
         this.webSocketConnectionProvider.listen({
-            path: `${terminalsPath}/${this.terminalId}`,
+            path: `${terminalsPath}/${this.termId}`,
```
I think for read-access it is fine to use `this.terminalId` not internal field.

akosyakov commented on packages/core/src/browser/widgets/widget.ts:
```diff
@@ -46,11 +46,14 @@ export class BaseWidget extends Widget {
         this.onScrollUpEmitter,
-        this.onDidChangeVisibilityEmitter
+        this.onDidChangeVisibilityEmitter,
+        this.onDidDisposeEmitter
```
It has to be first to be the last called. Thank you for adding it :)

akosyakov commented on packages/core/src/browser/widgets/widget.ts:
```diff
@@ -60,6 +63,7 @@ export class BaseWidget extends Widget {
         if (this.isDisposed) {
             return;
         }
+        this.onDidDisposeEmitter.fire();
```
I think it should go instead to `toDispose` collection. i.e.
```ts
protected readonly toDispose = new DisposableCollection(
        this.onDidDisposeEmitter,
        Disposable.create(() => this.onDidDisposeEmitter.fire(),
        this.onScrollYReachEndEmitter,
        this.onScrollUpEmitter,
        this.onDidChangeVisibilityEmitter,
    );
```

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    private getTaskTerminalWidgets(): TaskTerminalWidget[] {
+        const widgets: TaskTerminalWidget[] = [];
```
fyi: `return this.terminalService.all.filter(TaskTerminalWidget.is);`

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+
+        const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
+        this.terminalService.onDidCreateTerminal((widget: TerminalWidget) => {
+            const toDisposeCreateListener = widget.onTerminalDidConnectProcess(async (connectedWidget: TerminalWidget) => {
```
I think it is redundant to have `connectedWidget` and `widget`. There is not need for `connectedWidget`.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+                }
+                const toDisposeCloseEvent = connectedWidget.onDidDispose(() => {
+                    toDisposeCloseEvent.dispose();
+                    toDisposeCreateListener.dispose();
```
that's redundant, disposing emitters disposes all listeners

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+        });
+
+        const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
+        this.terminalService.onDidCreateTerminal((widget: TerminalWidget) => {
```
I think it should be properly guarded to be type-safe and don't mutate all terminals:
```ts
this.terminalService.onDidCreateTerminal(widget => {
    const terminal = TaskTerminalWidget.is(widget) && widget;
    if (terminal) {
        const didConectListener = terminal.onDidConnect(async () => {
        const tasksInfo = await this.taskServer.getTasks(context);
        const taskInfo = tasksInfo.find(info => info.terminalId === widget.terminalId);
        if (taskInfo) {
            const taskConfig = taskInfo.config;
            terminal.dedicated = !!taskConfig.presentation && !!taskConfig.presentation.panel && taskConfig.presentation.panel === PanelKind.Dedicated;
            terminal.taskId = taskInfo.taskId;
            terminal.taskConfig = taskConfig;
            terminal.busy = true;
        }
    });
    widget.onDidDispose(() => didConectListener.dispose());
}
```

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+                        busy: true
+                    });
+                }
+                const toDisposeCloseEvent = connectedWidget.onDidDispose(() => {
```
It has to go to outer if to dispose didConnect listener even if a terminal never connected

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+                        dedicated,
+                        taskId: taskInfo.taskId,
+                        taskConfig,
+                        busy: true
```
`busy` should be installed whenever associated task is running, it should not depend on `didConnect`. Frontend can lose and reestablish a connection several times while a task is running, `busy` flag should remain `true` all the time. I don't think you need  `onDidConnect` event.

elaihau replied:
i do need `onTerminalDidConnectProcess`. without it the line `const taskInfo = tasksInfo.find(info => info.terminalId === widget.terminalId);` will always assign `undefined` to `taskInfo`, and nothing in this if-block (i.e., if (taskInfo) { /* assign TaskTerminalWidget properties*/ }) will be run.

The reason behind is, before the `onTerminalDidConnectProcess` happens, `terminalWidget.terminalId` is -1. Please check the terminal-widget code for details.

akosyakov replied:
There is already `onDidOpen` event already which is fired then the terminal is started or reconnected.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,184 @@
+    dedicated?: boolean;
+    taskId?: number;
+    taskConfig?: TaskConfiguration;
+    busy?: boolean;
```
It would  be good to add js-doc of what you mean by busy looking at the PR we seem to have different understanding.

For me busy is whether associated task is running, it should be set when a task starts running and be unset when a tasks stop. It should not be changed by other events.
PR though sets it on `didConnect`, I don't understand why it is necessary and relevant.

elaihau replied:
I see why `busy` means same way as you do.
Please see the reason of using `onTerminalDidConnectProcess` from https://github.com/eclipse-theia/theia/pull/7260#discussion_r396827368

akosyakov commented:
@elaihau terminal API looks better, I'm still though with definition of busy for the task terminal and need for `didConnect` event. Please see https://github.com/eclipse-theia/theia/pull/7260#discussion_r396262615

akosyakov commented:
> I retested both scenarios you described. They should work now.

@elaihau Could you please create a task to add integration tests for terminals? We could make a first PR together that afterwards you can add more tests. But not as a part of this PR.

elaihau commented:
> task to add integration tests for terminals

https://github.com/eclipse-theia/theia/issues/7408 was created.

akosyakov commented on packages/core/src/browser/widgets/widget.ts:
```diff
@@ -46,11 +46,15 @@ export class BaseWidget extends Widget {
         this.onScrollUpEmitter,
-        this.onDidChangeVisibilityEmitter
+        this.onDidChangeVisibilityEmitter,
+        this.onDidDisposeEmitter,
```
these lines should be first to add to be run the last after all emitters and listeners are released

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,175 @@
+            }
+        });
+
+        const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
```
I think it should go just before `getTasks`, we should not cache it.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,175 @@
+                const didConectListener = terminal.onTerminalDidConnectProcess(async () => {
+                    const tasksInfo = await this.taskServer.getTasks(context);
+                    const taskInfo = tasksInfo.find(info => info.terminalId === widget.terminalId);
+                    if (taskInfo) {
```
I think here is `else` missing with `Terminal will be reused by tasks.`.

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -974,26 +978,30 @@ export class TaskService implements TaskConfigurationClient {
-                this.shell.revealWidget(widget.id);
-            }
-        }
         widget.start(processId);
```
this line worries me, `TaskTerminalWidgetManager` does not get any callback if start failed completely, so we will end up with something which looks like a user terminal

Do you know how to reproduce an issue when this method throws?

elaihau replied:
the way i can think of is making some code changes to the terminal extension to make sure the error throws.

akosyakov replied:
ok, it seems to be irrelevant if we do https://github.com/eclipse-theia/theia/pull/7260#discussion_r397760106 and handle it in the task terminal widget manager.

akosyakov commented on packages/task/src/browser/task-service.ts:
```diff
@@ -974,26 +978,30 @@ export class TaskService implements TaskConfigurationClient {
+        }
+
+        // Create / find a terminal widget to display an execution output of a task that was launched as a command inside a shell.
+        const widget = await this.taskTerminalWidgetManager.open(
```
Please file an issue to separate identity from opener options for terminals and task terminals in particularly. We put too much stuff into the identity options, i.e.  `widgetOptions`, `mode`, `taskConfig` should not be stored. Local storage is not big just 5 mb for all origin data.

elaihau replied:
i will create the GH issue after this PR gets merged.

elaihau commented:
I changed the terminal extension to implemented what you suggested @akosyakov 

Please kindly let me knwo if this is what you think the terminal should do. Thanks !

![Peek 2020-03-24 22-00](https://user-images.githubusercontent.com/37082801/77494001-fa8d7100-6e1a-11ea-870a-04ea817fc4c5.gif)

akosyakov commented:
@elaihau Yes, it is working how i would imagine it! The task terminal remains task terminal after reload and can be reused.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,182 @@
+        this.terminalService.onDidCreateTerminal(async (widget: TerminalWidget) => {
+            const terminal = TaskTerminalWidget.is(widget) && widget;
+            if (terminal) {
+                const didConectListener = terminal.onDidOpen(async () => {
```
`Conect` -> `Connect`

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,182 @@
+                        terminal.taskConfig = taskConfig;
+                        terminal.busy = true;
+                    } else {
+                        terminal.busy = false;
```
`terminal.busy = false` can go into `notifyTaskFinished`?

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -310,11 +322,11 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+        this._terminalId = typeof id !== 'number' ? await this.createTerminal() : await this.attachTerminal(id);
         this.resizeTerminalProcess();
         this.connectTerminalProcess();
+        this.onDidOpenEmitter.fire(undefined);
```
It should fire only if start was successful, please move it back.

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,182 @@
+            const terminal = TaskTerminalWidget.is(widget) && widget;
+            if (terminal) {
+                const didConectListener = terminal.onDidOpen(async () => {
+                    const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
```
Should not we dispose `didConnectListener` here? i.e. intermediate reconnections should not affect busy state.

We should test it by:
- creating a new task terminal
- running a long running task
- switching off connection to the server
- trying to run a new long running task
- switching on connection to the server
- the original task terminal should not be used for second task, whether should not be any unexpected effects, after reconnection is reestablished, the output of original long running task should be shown.

It is important the server should not be stopped and started. It is different use case. You can use Gitpod and switch off/no your internet connection to reproduce or some local proxy to drop traffic to localhost.

elaihau replied:
> * switching off connection to the server
> * trying to run a new long running task

sorry one thing confuses me: how am i able to start a new long running task while the connection is off?

elaihau replied:
> Should not we dispose `didConnectListener` here? i.e. intermediate reconnections should not affect busy state.

I tested the scenarios of server reconnects on Gitpod.

Server reconnect works - terminal widget successfully attaches to the process, task output is properly displayed in the terminal widget, and the terminal widget can be used for the next task.
However I believe `didConnectListener` should not be disposed in the callback of `terminal.onDidOpen`. It causes `Terminal will be reused by tasks.` not to be displayed for new tasks that are started after the reconnect.

akosyakov replied:
ok, to me most important that you test such cases :) If code is required to make them work let's keep it

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -310,11 +322,11 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
-            this.onDidOpenEmitter.fire(undefined);
             return this.terminalId;
         }
         throw new Error('Failed to start terminal' + (id ? ` for id: ${id}.` : '.'));
```
let's fire a new event here `onDidOpenFailure`, you will need to wait for it the task widget manager

elaihau commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -326,7 +338,7 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
         this.logger.error(`Error attaching to terminal id ${id}, the terminal is most likely gone. Starting up a new terminal instead.`);
-        return this.createTerminal();
+        return -1;
```
yep it works as you described.

![Peek 2020-03-25 18-29](https://user-images.githubusercontent.com/37082801/77595073-94622600-6ece-11ea-8496-52fda3c9cfc5.gif)

akosyakov replied:
cool, thanks for checking 👍

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+export namespace TaskTerminalWidgetOpenerOptions {
```
(nit) I will put it next to interface

akosyakov commented on packages/task/src/browser/task-terminal-widget-manager.ts:
```diff
@@ -0,0 +1,185 @@
+                        terminal.taskId = taskInfo.taskId;
+                        terminal.taskConfig = taskConfig;
+                        terminal.busy = true;
+                    }
```
are you sure that `else` is not required? this code is async - could it be that a task was finished in the meantime?

elaihau replied:
you are right. i added the else block back in

RomanNikitenko commented:
Testing right now, will report the result here

marcdumais-work commented:
@akosyakov @vince-fugnitto @RomanNikitenko What's your feeling about merging this just before the release vs waiting after? (Assuming Roman reports positively after last round of tests and there is time)

vince-fugnitto commented:
> @akosyakov @vince-fugnitto @RomanNikitenko What's your feeling about merging this just before the release vs waiting after? (Assuming Roman reports positively after last round of tests and there is time)

@marcdumais-work I'm fine either way, if we want to be extra cautious I guess we can hold off on merging till right after the release. It's unfortunate since the changes are quite nice.

lmcbout commented:
What about waiting another ~3 hours to see the comments from @RomanNikitenko before starting the release !

akosyakov commented:
I don't have strong opinion about it. Is there an urgent reason to land it before 1.0?

marcdumais-work commented:
> Is there an urgent reason to land it before 1.0?

I do not think so, just a "nice to have". So if there is even a little bit of perceived risk, maybe we can wait after.

akosyakov commented:
> I do not think so, just a "nice to have". So if there is even a little bit of perceived risk, maybe we can wait after.

ok, let's wait we don't reset terminals anybody on reconnect. It does not seem to break during testing, but maybe we overlooked something. Let's merge after the release and we will have the whole month to learnt :)

elaihau commented:
Thank you all for your time and efforts !

vince-fugnitto commented:
> Thank you all for your time and efforts !

@elaihau I think you can merge now that the release has been completed :)
